### PR TITLE
feat: process and map dedicated device.* fields

### PR DESCRIPTION
**Pull request by kruskall**
## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues

Closes https://github.com/elastic/apm-server/issues/8957

### CODE REVIEW

File: internal/model/device.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	ID string
+
+	// Architecture holds the host machine architecture.
+	Model Model
+
+	// Manufacturer the vendor name of the device manufacturer.
+	Manufacturer string
+}
+
+type Model struct {
```
Comment (axw) [inline internal/model/device.go:35]:
```suggestion
type DeviceModel struct {
```

Calling this DeviceModel will make the name a bit more meaningful when looking at it out of the context of this file, e.g. by looking at godoc.

File: internal/model/device.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+type Device struct {
+	// ID holds the unique identifier of a device.
+	ID string
+
+	// Architecture holds the host machine architecture.
```
Comment (axw) [inline internal/model/device.go:28]:
```suggestion
	// Model holds information about the device model.
```

File: internal/model/device.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	Name string
+
+	// Identifier holds the machine readable identifier of the device model.
+	Identifier string
+}
+
+func (d *Device) fields() mapstr.M {
+	if d == nil {
+		return nil
+	}
```
Comment (axw) [inline internal/model/device.go:46]:
```suggestion
	if d == nil {
		return nil
	}
```

Do we need it? APMEvent.Device is not a pointer.

**Review by axw — CHANGES_REQUESTED** [review]
Looks great! We should also add these fields to `ecs.yml` in `app_logs`, `app_metrics`, `error_logs`, and `traces` data streams.

I forgot about that for the previous PR, so can you please also add the fields added in https://github.com/elastic/apm-server/pull/8784/files while you're at it?

File: internal/model/device.go (hunk 1 above)
Comment (kruskall) [reply internal/model/device.go]: Good point! Updated!

File: internal/model/device.go (hunk 2 above)
Comment (kruskall) [reply internal/model/device.go]: Thank you for spotting this! 🙇

File: internal/model/device.go (hunk 3 above)
Comment (kruskall) [reply internal/model/device.go]: I couldn't see a consistent usage of nil checks in other `fields` methods. I guess it's fine to remove it.

**Comment by kruskall** [comment]
> ooks great! We should also add these fields to `ecs.yml` in `app_logs`, `app_metrics`, `error_logs`, and `traces` data streams.
> 
> I forgot about that for the previous PR, so can you please also add the fields added in https://github.com/elastic/apm-server/pull/8784/files while you're at it?

@axw sory, I'm a bit lost here 😓 

`build/apmpackage` is failing but the definition in `ecs.yml` seems to be correct. I see the apm-server is using an older version of https://github.com/elastic/elastic-package/ but updating yields the same result.

**Comment by axw** [comment]
@kruskall I think the issue is that `device.*` fields aren't in any released version of ECS. We have a couple of options:
 - update https://github.com/elastic/apm-server/blob/bd0969db37e6f6200b496cf2fa10b3d3239323ff/apmpackage/apm/_dev/build/build.yml#L4 to use `git@main`
 - write out those the fields by hand

I think it's probably safe to use the main branch. Maybe try that, and diff the build package before and after, to check that none of the fields types have changed incompatibly.

**Comment by kruskall** [comment]
> I think the issue is that `device.*` fields aren't in any released version of ECS. We have a couple of options:
> 
>     * update https://github.com/elastic/apm-server/blob/bd0969db37e6f6200b496cf2fa10b3d3239323ff/apmpackage/apm/_dev/build/build.yml#L4

Thank you for the help and guidance! 🙇 Seems to be fixed!

**Review by axw — APPROVED** [review]
LGTM!

**Pull request merged**